GitUI/UserControls/RevisionGrid/Graph/RevisionGraphRevision.cs

```diff
@@ -14,12 +14,13 @@ namespace GitUI.UserControls.RevisionGrid.Graph
     //     *  <- parent revision
     public class RevisionGraphRevision
     {
+        private ImmutableStack<RevisionGraphRevision> _parents = ImmutableStack<RevisionGraphRevision>.Empty;
```

Comment:
💭  I believe an ideal collection here would allow up to 2 `RevisionGraphRevision` parents without allocations. For example, it could be a struct type with an `object` field, where the field has type `RevisionGraphRevision` for a single parent, or `RevisionGraphRevision[]` for more than 2 parents.

```csharp
struct RevisionGraphRelativesCollection
{
  private object _singleOrAllRelatives;
  private RevisionGraphRevision _secondRelativeIfExactlyTwo;
}
```

The current `ImmutableStack<T>` performed well enough to use it as a clean incremental improvement.

Reply:
I looked at the source of ImmutableStack, its pretty much implemented to do this. I didn't know about this type before but it seems pretty fast. I am only curious about the performance of the indexer, but that is never used for parents/children/segments so we are ok here.